File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/dataview/StateMapView.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.flink.table.dataview

import java.util
import java.lang.{Iterable => JIterable}
import java.util.{Iterator => JIterator, Map => JMap}

import org.apache.flink.runtime.state.subkeyed.{SubKeyedMapState, SubKeyedValueState}
import org.apache.flink.runtime.state.keyed.{KeyedMapState, KeyedValueState}
import org.apache.flink.table.api.dataview.MapView
import org.apache.flink.table.dataview.NullAwareMapIterator.NullMapEntry

// -------------------------------------------------------------------------------------
//                                State MapView
// -------------------------------------------------------------------------------------

trait StateMapView[K, MK, MV] extends MapView[MK, MV] with StateDataView[K]

/**
  * [[SubKeyedStateMapView]] is a [[SubKeyedMapState]] with [[MapView]] interface which works on
  * window aggregate.
  */
class SubKeyedStateMapView[K, N, MK, MV](state: SubKeyedMapState[K, N, MK, MV])
  extends StateMapView[K, MK, MV] {

  private var key: K = _
  private var namespace: N = _

  override def setCurrentKey(key: K): Unit = {
    this.key = key
  }

  def setCurrentNamespace(namespace: N): Unit = {
    this.namespace = namespace
  }

  override def get(mapKey: MK): MV = state.get(key, namespace, mapKey)

  override def put(mapKey: MK, mapValue: MV): Unit = state.add(key, namespace, mapKey, mapValue)

  override def putAll(map: JMap[MK, MV]): Unit = state.addAll(key, namespace, map)

  override def remove(mapKey: MK): Unit = state.remove(key, namespace, mapKey)

  override def contains(mapKey: MK): Boolean = state.contains(key, namespace, mapKey)

  override def entries: JIterable[JMap.Entry[MK, MV]] = {
    new JIterable[JMap.Entry[MK, MV]]() {
      override def iterator(): util.Iterator[JMap.Entry[MK, MV]] = state.iterator(key, namespace)
    }
  }

  override def keys: JIterable[MK] = new JIterable[MK]() {
    override def iterator(): util.Iterator[MK] = new util.Iterator[MK] {
      val iter: util.Iterator[JMap.Entry[MK, MV]] = state.iterator(key, namespace)

      override def next(): MK = iter.next().getKey

      override def hasNext: Boolean = iter.hasNext
    }
  }

  override def values: JIterable[MV] = new JIterable[MV]() {
    override def iterator(): util.Iterator[MV] = new util.Iterator[MV] {
      val iter: util.Iterator[JMap.Entry[MK, MV]] = state.iterator(key, namespace)

      override def next(): MV = iter.next().getValue

      override def hasNext: Boolean = iter.hasNext
    }
  }

  override def iterator: util.Iterator[JMap.Entry[MK, MV]] = state.iterator(key, namespace)

  override def clear(): Unit = state.remove(key, namespace)
}

/**
  * [[KeyedStateMapView]] is a [[KeyedMapState]] with [[MapView]] interface which works on
  * group aggregate.
  */
class KeyedStateMapView[K, MK, MV](state: KeyedMapState[K, MK, MV])
  extends StateMapView[K, MK, MV] {

  private var stateKey: K = null.asInstanceOf[K]

  override def setCurrentKey(key: K): Unit = {
    this.stateKey = key
  }

  override def get(key: MK): MV = {
    state.get(stateKey, key)
  }

  override def put(key: MK, value: MV): Unit = {
    state.add(stateKey, key, value)
  }

  override def putAll(map: JMap[MK, MV]): Unit = {
    state.addAll(stateKey, map)
  }

  override def remove(key: MK): Unit = {
    state.remove(stateKey, key)
  }

  override def contains(key: MK): Boolean = {
    state.contains(stateKey, key)
  }

  override def entries: JIterable[JMap.Entry[MK, MV]] = {
    new JIterable[JMap.Entry[MK, MV]] {
      override def iterator(): util.Iterator[JMap.Entry[MK, MV]] = state.iterator(stateKey)
    }
  }

  override def keys: JIterable[MK] = {
    new JIterable[MK] {
      override def iterator(): util.Iterator[MK] = new util.Iterator[MK] {
        val it: JIterator[JMap.Entry[MK, MV]] = state.iterator(stateKey)
        override def next(): MK = it.next().getKey
        override def hasNext: Boolean = it.hasNext
      }
    }
  }

  override def values: JIterable[MV] = {
    new JIterable[MV] {
      override def iterator(): util.Iterator[MV] = new util.Iterator[MV] {
        val it: JIterator[JMap.Entry[MK, MV]] = state.iterator(stateKey)
        override def next(): MV = it.next().getValue
        override def hasNext: Boolean = it.hasNext
      }
    }
  }

  override def iterator: util.Iterator[JMap.Entry[MK, MV]] = {
    state.iterator(stateKey)
  }

  override def clear(): Unit = state.remove(stateKey)
}

// -------------------------------------------------------------------------------------
//                                NullAware State MapView
// -------------------------------------------------------------------------------------


/**
  * [[NullAwareSubKeyedStateMapView]] is an implementation of [[MapView]] using
  * [[SubKeyedMapState]] and [[SubKeyedValueState]] which can handle null map key.
  */
class NullAwareSubKeyedStateMapView[K, N, MK, MV](
    mapState: SubKeyedMapState[K, N, MK, MV],
    nullState: SubKeyedValueState[K, N, MV])
  extends StateMapView[K, MK, MV] {

  private var key: K = _
  private var namespace: N = _

  override def setCurrentKey(key: K): Unit = {
    this.key = key
  }

  def setCurrentNamespace(namespace: N): Unit = {
    this.namespace = namespace
  }

  override def get(mapKey: MK): MV = {
    if (mapKey == null) {
      nullState.get(key, namespace)
    } else {
      mapState.get(key, namespace, mapKey)
    }
  }

  override def put(mapKey: MK, mapValue: MV): Unit = {
    if (mapKey == null) {
      nullState.put(key, namespace, mapValue)
    } else {
      mapState.add(key, namespace, mapKey, mapValue)
    }
  }

  override def putAll(map: JMap[MK, MV]): Unit = {
    val iter = map.entrySet().iterator()
    while (iter.hasNext) {
      val entry = iter.next()
      val mapKey = entry.getKey
      val mapValue = entry.getValue
      put(mapKey, mapValue)
    }
  }

  override def remove(mapKey: MK): Unit = {
    if (mapKey == null) {
      nullState.remove(key, namespace)
    } else {
      mapState.remove(key, namespace, mapKey)
    }
  }

  override def contains(mapKey: MK): Boolean = {
    if (mapKey == null) {
      nullState.contains(key, namespace)
    } else {
      mapState.contains(key, namespace, mapKey)
    }
  }

  override def entries: JIterable[JMap.Entry[MK, MV]] = {
    new JIterable[JMap.Entry[MK, MV]]() {
      override def iterator(): util.Iterator[JMap.Entry[MK, MV]] =
        NullAwareSubKeyedStateMapView.this.iterator
    }
  }

  override def keys: JIterable[MK] = new JIterable[MK]() {
    override def iterator(): util.Iterator[MK] = new util.Iterator[MK] {
      val iter: util.Iterator[JMap.Entry[MK, MV]] = NullAwareSubKeyedStateMapView.this.iterator

      override def next(): MK = iter.next().getKey

      override def hasNext: Boolean = iter.hasNext
    }
  }

  override def values: JIterable[MV] = new JIterable[MV]() {
    override def iterator(): util.Iterator[MV] = new util.Iterator[MV] {
      val iter: util.Iterator[JMap.Entry[MK, MV]] = NullAwareSubKeyedStateMapView.this.iterator

      override def next(): MV = iter.next().getValue

      override def hasNext: Boolean = iter.hasNext
    }
  }

  override def iterator: util.Iterator[JMap.Entry[MK, MV]] = {
    new NullAwareMapIterator[MK, MV](
      mapState.iterator(key, namespace),
      new NullMapEntry[MK, MV] {
        override def remove(): Unit = nullState.remove(key, namespace)

        override def getValue: MV = nullState.get(key, namespace)

        override def setValue(value: MV): MV = {
          nullState.put(key, namespace, value)
          value
        }
      })
  }

  override def clear(): Unit = {
    mapState.remove(key, namespace)
    nullState.remove(key, namespace)
  }
}

/**
  * [[NullAwareKeyedStateMapView]] is an implementation of [[MapView]] using [[KeyedStateMapView]]
  * and [[KeyedValueState]] which can handle null map key.
  */
class NullAwareKeyedStateMapView[K, MK, MV](
    mapState: KeyedMapState[K, MK, MV],
    nullState: KeyedValueState[K, MV])
  extends StateMapView[K, MK, MV] {

  private var stateKey: K = null.asInstanceOf[K]

  override def setCurrentKey(key: K): Unit = {
    this.stateKey = key
  }

  override def get(key: MK): MV = {
    if (key == null) {
      nullState.get(stateKey)
    } else {
      mapState.get(stateKey, key)
    }
  }

  override def put(key: MK, value: MV): Unit = {
    if (key == null) {
      nullState.put(stateKey, value)
    } else {
      mapState.add(stateKey, key, value)
    }
  }

  override def putAll(map: JMap[MK, MV]): Unit = {
    val iter = map.entrySet().iterator()
    while (iter.hasNext) {
      val entry = iter.next()
      val key = entry.getKey
      val value = entry.getValue
      put(key, value)
    }
  }

  override def remove(key: MK): Unit = {
    if (key == null) {
      nullState.remove(stateKey)
    } else {
      mapState.remove(stateKey, key)
    }
  }

  override def contains(key: MK): Boolean = {
    if (key == null) {
      nullState.contains(stateKey)
    } else {
      mapState.contains(stateKey, key)
    }
  }

  override def entries: JIterable[JMap.Entry[MK, MV]] = {
    new JIterable[JMap.Entry[MK, MV]] {
      override def iterator(): util.Iterator[JMap.Entry[MK, MV]] =
        NullAwareKeyedStateMapView.this.iterator
    }
  }

  override def keys: JIterable[MK] = {
    new JIterable[MK] {
      override def iterator(): util.Iterator[MK] = new util.Iterator[MK] {
        val it: JIterator[JMap.Entry[MK, MV]] = NullAwareKeyedStateMapView.this.iterator
        override def next(): MK = it.next().getKey
        override def hasNext: Boolean = it.hasNext
      }
    }
  }

  override def values: JIterable[MV] = {
    new JIterable[MV] {
      override def iterator(): util.Iterator[MV] = new util.Iterator[MV] {
        val it: JIterator[JMap.Entry[MK, MV]] = NullAwareKeyedStateMapView.this.iterator
        override def next(): MV = it.next().getValue
        override def hasNext: Boolean = it.hasNext
      }
    }
  }

  override def iterator: util.Iterator[JMap.Entry[MK, MV]] = {
    new NullAwareMapIterator[MK, MV](
      mapState.iterator(stateKey),
      new NullMapEntry[MK, MV] {
        override def remove(): Unit = nullState.remove(stateKey)

        override def getValue: MV = nullState.get(stateKey)

        override def setValue(value: MV): MV = {
          nullState.put(stateKey, value)
          value
        }
      })
  }

  override def clear(): Unit = {
    mapState.remove(stateKey)
    nullState.remove(stateKey)
  }
}

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataview/StateMapView.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.dataview;

import org.apache.flink.api.common.state.MapState;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.runtime.state.internal.InternalMapState;
import org.apache.flink.runtime.state.internal.InternalValueState;
import org.apache.flink.table.api.dataview.MapView;
import org.apache.flink.util.IterableIterator;

import java.io.IOException;
import java.util.Collections;
import java.util.Iterator;
import java.util.Map;

/**
 * MapView which is implemented using state backends.
 *
 * @param <N> the type of namespace
 * @param <MK> the type of MapView key
 * @param <MV> the type of MapView value
 */
public abstract class StateMapView<N, MK, MV> extends MapView<MK, MV> implements StateDataView<N> {

	private static final long serialVersionUID = 1L;

	/**
	 * {@link StateMapViewWithKeysNotNull} is a {@link MapView} which implemented
	 * using state backend which map keys shouldn't be null. This is the default
	 * implementation for {@link StateMapView}.
	 *
	 * @param <N> the type of namespace
	 * @param <MK> the type of MapView key
	 * @param <MV> the type of MapView value
	 */
	private abstract static class StateMapViewWithKeysNotNull<N, MK, MV> extends StateMapView<N, MK, MV> {

		private static final long serialVersionUID = 2605280027745112384L;

		private final MapState<MK, MV> mapState;
		private final Map<MK, MV> emptyState = Collections.emptyMap();

		private StateMapViewWithKeysNotNull(MapState<MK, MV> mapState) {
			this.mapState = mapState;
		}

		@Override
		public MV get(MK key) throws Exception {
			return mapState.get(key);
		}

		@Override
		public void put(MK key, MV value) throws Exception {
			mapState.put(key, value);
		}

		@Override
		public void putAll(Map<MK, MV> map) throws Exception {
			mapState.putAll(map);
		}

		@Override
		public void remove(MK key) throws Exception {
			mapState.remove(key);
		}

		@Override
		public boolean contains(MK key) throws Exception {
			return mapState.contains(key);
		}

		@Override
		public Iterable<Map.Entry<MK, MV>> entries() throws Exception {
			Iterable<Map.Entry<MK, MV>> original = mapState.entries();
			return original != null ? original : emptyState.entrySet();
		}

		@Override
		public Iterable<MK> keys() throws Exception {
			Iterable<MK> original = mapState.keys();
			return original != null ? original : emptyState.keySet();
		}

		@Override
		public Iterable<MV> values() throws Exception {
			Iterable<MV> original = mapState.values();
			return original != null ? original : emptyState.values();
		}

		@Override
		public Iterator<Map.Entry<MK, MV>> iterator() throws Exception {
			Iterator<Map.Entry<MK, MV>> original = mapState.iterator();
			return original != null ? original : emptyState.entrySet().iterator();
		}

		@Override
		public void clear() {
			mapState.clear();
		}
	}

	/**
	 * {@link StateMapViewWithKeysNullable} is a {@link MapView} which implemented using state backend
	 * and can handle nullable map keys. Currently this is only used internally when implementing
	 * distinct aggregates.
	 *
	 * @param <N> the type of namespace
	 * @param <MK> the type of MapView key
	 * @param <MV> the type of MapView value
	 */
	private abstract static class StateMapViewWithKeysNullable<N, MK, MV> extends StateMapView<N, MK, MV> {

		private static final long serialVersionUID = 2605280027745112384L;

		private final MapState<MK, MV> mapState;
		private final ValueState<MV> nullState;

		private StateMapViewWithKeysNullable(
			MapState<MK, MV> mapState,
			ValueState<MV> nullState) {

			this.mapState = mapState;
			this.nullState = nullState;
		}

		@Override
		public MV get(MK key) throws Exception {
			if (key == null) {
				return nullState.value();
			} else {
				return mapState.get(key);
			}
		}

		@Override
		public void put(MK key, MV value) throws Exception {
			if (key == null) {
				nullState.update(value);
			} else {
				mapState.put(key, value);
			}
		}

		@Override
		public void putAll(Map<MK, MV> map) throws Exception {
			for (Map.Entry<MK, MV> entry : map.entrySet()) {
				// entry key might be null, so we can't invoke mapState.putAll(map) directly here
				put(entry.getKey(), entry.getValue());
			}
		}

		@Override
		public void remove(MK key) throws Exception {
			if (key == null) {
				nullState.clear();
			} else {
				mapState.remove(key);
			}
		}

		@Override
		public boolean contains(MK key) throws Exception {
			if (key == null) {
				return nullState.value() != null;
			} else {
				return mapState.contains(key);
			}
		}

		@Override
		public Iterable<Map.Entry<MK, MV>> entries() throws Exception {
			final Iterator<Map.Entry<MK, MV>> iterator = iterator();
			return () -> iterator;
		}

		@Override
		public Iterable<MK> keys() throws Exception {
			return new KeysIterable(this.iterator());
		}

		@Override
		public Iterable<MV> values() throws Exception {
			return new ValuesIterable(this.iterator());
		}

		@Override
		public Iterator<Map.Entry<MK, MV>> iterator() throws Exception {
			return new NullAwareMapIterator<>(mapState.iterator(), new NullMapEntryImpl());
		}

		@Override
		public void clear() {
			mapState.clear();
			nullState.clear();
		}

		/**
		 * MapEntry for the null key of this MapView.
		 */
		private class NullMapEntryImpl implements NullAwareMapIterator.NullMapEntry<MK, MV> {

			@Override
			public MV getValue() {
				try {
					return nullState.value();
				} catch (IOException e) {
					throw new RuntimeException(e);
				}
			}

			@Override
			public MV setValue(MV value) {
				MV oldValue;
				try {
					oldValue = nullState.value();
					nullState.update(value);
					return oldValue;
				} catch (IOException e) {
					throw new RuntimeException(e);
				}
			}

			@Override
			public void remove() {
				nullState.clear();
			}
		}

		/**
		 * Iterable of the keys on the MapView.
		 */
		private class KeysIterable implements IterableIterator<MK> {

			private final Iterator<Map.Entry<MK, MV>> iterator;

			private KeysIterable(Iterator<Map.Entry<MK, MV>> iterator) {
				this.iterator = iterator;
			}

			@Override
			public Iterator<MK> iterator() {
				return this;
			}

			@Override
			public boolean hasNext() {
				return iterator.hasNext();
			}

			@Override
			public MK next() {
				return iterator.next().getKey();
			}
		}

		/**
		 * Iterable of the values on the MapView.
		 */
		private class ValuesIterable implements IterableIterator<MV> {

			private final Iterator<Map.Entry<MK, MV>> iterator;

			private ValuesIterable(Iterator<Map.Entry<MK, MV>> iterator) {
				this.iterator = iterator;
			}

			@Override
			public Iterator<MV> iterator() {
				return this;
			}

			@Override
			public boolean hasNext() {
				return iterator.hasNext();
			}

			@Override
			public MV next() {
				return iterator.next().getValue();
			}
		}
	}

	/**
	 * A state MapView which do not support nullable keys and namespace.
	 */
	public static final class KeyedStateMapViewWithKeysNotNull<N, MK, MV> extends StateMapViewWithKeysNotNull<N, MK, MV> {

		private static final long serialVersionUID = 6650061094951931356L;

		public KeyedStateMapViewWithKeysNotNull(MapState<MK, MV> mapState) {
			super(mapState);
		}

		@Override
		public void setCurrentNamespace(N namespace) {
			throw new UnsupportedOperationException();
		}
	}

	/**
	 * A state MapView which support namespace but not support nullable keys.
	 */
	public static final class NamespacedStateMapViewWithKeysNotNull<N, MK, MV> extends StateMapViewWithKeysNotNull<N, MK, MV> {

		private static final long serialVersionUID = -2793150592169689571L;
		private final InternalMapState<?, N, MK, MV> internalMapState;

		public NamespacedStateMapViewWithKeysNotNull(InternalMapState<?, N, MK, MV> internalMapState) {
			super(internalMapState);
			this.internalMapState = internalMapState;
		}

		@Override
		public void setCurrentNamespace(N namespace) {
			internalMapState.setCurrentNamespace(namespace);
		}
	}

	/**
	 * A state MapView which support nullable keys, but not support namespace.
	 */
	public static final class KeyedStateMapViewWithKeysNullable<N, MK, MV> extends StateMapViewWithKeysNullable<N, MK, MV> {

		private static final long serialVersionUID = -4222930534937318207L;

		public KeyedStateMapViewWithKeysNullable(MapState<MK, MV> mapState, ValueState<MV> nullState) {
			super(mapState, nullState);
		}

		@Override
		public void setCurrentNamespace(N namespace) {
			throw new UnsupportedOperationException();
		}
	}

	/**
	 * A state MapView which support nullable keys and namespace.
	 */
	public static final class NamespacedStateMapViewWithKeysNullable<N, MK, MV> extends StateMapViewWithKeysNullable<N, MK, MV> {

		private static final long serialVersionUID = -6915428707804508152L;
		private final InternalMapState<?, N, MK, MV> internalMapState;
		private final InternalValueState<?, N, MV> internalNullState;

		public NamespacedStateMapViewWithKeysNullable(InternalMapState<?, N, MK, MV> internalMapState, InternalValueState<?, N, MV> internalNullState) {
			super(internalMapState, internalNullState);
			this.internalMapState = internalMapState;
			this.internalNullState = internalNullState;
		}

		@Override
		public void setCurrentNamespace(N namespace) {
			internalMapState.setCurrentNamespace(namespace);
			internalNullState.setCurrentNamespace(namespace);
		}
	}

}
